### ISSUE TITLE
Design review and discussion: MFLUX v1.0 CLI

### ISSUE THREAD

**Issue by filipstrand** [filipstrand/mflux#357]
As the project has evolved and now support a whole family of models and techniques, the current CLI starts to feel outdated and unnecessarily complex. This is a place to discuss how a new and modern interface could/should look like given what we know about model families, variants and features in 2026 and beyond. 

- Which top level commands should we support (like `mflux generate`, `mflux edit`, `mflux save` etc?)
- What are good names (e.g --image or --images might be preferred to "--image_paths")?
- How should model families be specified? 
- How should model variants be specified (4b, 7b)? 
- How should LoRAs be specified (some cleanups and redesign have already been made here, e.g how we fetch from HF where the repo might include several .safetensors files etc). 
- How should resolution be specified or inferred (via input image, via aspect ratio, default values etc?) 
- How should low-ram optimisations be specified (e.g, what should be hidden behind a single --low-ram vs treated as sepearate, e.g see. https://github.com/filipstrand/mflux/pull/351) 
- what is natural to have to specify given that we already have an input image in some case (e.g default resolution could be inferred if not specified)
- anticipate/consider what could be coming in the future (e.g more optimisations, other model types etc). 
- how do we document this for humans?
- how do we document this for agents?
- many more things...

I invite everyone to participate in this open discussion, there are no right or wrong answers :) 

I think the more we can share how we use the project currently, and how we wish things to be, the better shot we have at designing a stable and good CLI for the future.  

Once we agree on a design, I think the actual implementation should be fairly trivial with the help of coding agents. 

I'll update this doc with some of my own suggestions after finishing current ongoing work with model porting...

**Comment by azrahello** [comment]
Hi again,
following up on our previous exchange — wanted to bring up a few features that I think could make a real difference in day-to-day mflux workflows.
1. API server support
With so many inference servers emerging (oMLX, MLX-LM server, and others), having a proper API layer in mflux could be a real game changer — not just for scripting, but for creative integrations: chaining LLM/VLM and T2I models together in pipelines, connecting to node-based UIs, or using tools like Claude Code as a coding agent against a local backend. An OpenAI-compatible endpoint would make mflux a first-class citizen in this growing ecosystem.
2. Shared checkpoints / component reuse
When downloading multiple fine-tuned or merged variants of Z-Image-Turbo (or other models), each comes with its own copy of the text encoder(s) and VAE — even when they're identical. On a system where storage matters, this adds up fast. It would be great to support shared checkpoints, similar to how ComfyUI handles model components: load the VAE and text encoder once, swap only the transformer weights.
3. Config files and environment-level defaults
mflux already has some config support, which is great — but there's room to expand it. For example, being able to set a default text encoder path or VAE path as a persistent default (config file or env var) would avoid redundant CLI arguments across sessions, especially when combining shared checkpoints as described above. Small ergonomic improvement, big daily workflow impact.
4. LoKR adapter support
I've been running into models that use LoKR (Low-rank Kronecker product) adapters and can't get them to work with mflux. As more community fine-tunes use LoKR (it's increasingly common in certain training frameworks), native support would help a lot. Happy to share specific model examples if that helps narrow down the implementation scope.
5. Image-to-image and cross-model blending
One thing I find limiting is the lack of tight image-to-image integration within mflux itself. For example, being able to pipe the output of one model as input to another — something like zit2klein — would be really useful when working with specialized or merged models that handle the same task differently. Instead of choosing one model, you could take the best of both worlds: generate with model A, refine or reinterpret with model B, all within the same tool without jumping to external pipelines.
6. Metadata embedding
I find metadata support extremely useful and would love to see it expanded. A few reasons:

JSON metadata is trivially scriptable — stripping a prompt from a JSON sidecar or embedded field takes one line, which makes it very easy to build training datasets directly from generated images.
While having the prompt visible in Preview or Lightroom is convenient, it's not always practical: Lightroom is heavy to keep open just for this, and Preview displays long prompts in blocks, requiring a separate copy-paste for each chunk. Having the full prompt in a JSON structure makes retrieval immediate and frictionless regardless of the tool you're using.
Would love to hear thoughts from maintainers and other users. Happy to test any of these if work starts on them.

@waldheinz I saw your notes on the training-related improvements you've been prototyping locally, specifically plotting training loss directly instead of the expensive validation loss computation, and gradient accumulation support. Both sound really solid, especially the loss plot optimization — spending 3/4 of compute on validation loss for samples that aren't even held out is a real pain point. If you're planning to open PRs for either or both, I'd be very happy to test them on my end (M2 Ultra, 64GB RAM). Just let me know.

**Comment by ianscrivener** [comment]
**bravo** @azrahello 🏆

Strongly prefer multiple LoRA arguments for multiple LoRAs `-- loraA 1.0 --loraB 0.5`

I LOVE `--from-config`. Guessing that CLI argument overrided config?

Currently coding Agent skills for DrawThings and other image inference engines.

**Comment by ianscrivener** [comment]
**`--from-config` - cascading**

I can see enormous benefit in having **multiple** `--from-config` . Consider;
```
 --from-config verbose_debugging.toml \
 --from-config unicorn_test.toml 
``` 

**Control output / debugging / verbose with `--from-config`** 
It would be great to be able to control the response output - say you were using Prometheus or something to pick up your CLI outputs.

**Comment by fxd0h** [comment]
@waldheinz picked up your point about the loss plot recomputing a validation forward every plot step over the same samples (no held-out set). implemented your suggestion to plot the free training loss instead: [972718b](https://github.com/fxd0h/mflux/commit/972718b). on a z-image run at plot_frequency=1 the per-step validation forward is gone (get_validation_batch calls went 6 to 1, just the initial baseline) and loss is still plotted each step. happy to open a PR if useful.

**Comment by fxd0h** [comment]
One thing worth folding into the v1.0 CLI design: a machine-readable dump of which options each model actually honours. Downstream tools transcribe capability tables from `--help` today, and `--help` overstates: the shared parser exposes `--negative-prompt` on every image CLI while FLUX.1 discards it unused and Ideogram 4 has no parameter for it (details and line refs in #498). `ModelConfig` already carries `supports_guidance`, so part of the data exists as structured state; something like `mflux capabilities --json`, or a per-model section emitted from that same source, would let tools stop guessing. It also fits the agent-facing documentation direction discussed here: an agent can read a JSON contract far more reliably than prose help text.

**Comment by ianscrivener** [comment]
Agree. An additional use case for this is to feed into Agent SKILLS/MCP

> One thing worth folding into the v1.0 CLI design: a machine-readable dump of which options each model actually honours. Downstream tools transcribe capability tables from `--help` today, and `--help` overstates: the shared parser exposes `--negative-prompt` on every image CLI while FLUX.1 discards it unused and Ideogram 4 has no parameter for it (details and line refs in [#498](https://github.com/filipstrand/mflux/issues/498)). `ModelConfig` already carries `supports_guidance`, so part of the data exists as structured state; something like `mflux capabilities --json`, or a per-model section emitted from that same source, would let tools stop guessing. It also fits the agent-facing documentation direction discussed here: an agent can read a JSON contract far more reliably than prose help text.

**Comment by ianscrivener** [comment]
**LLM.txt** also perhaps should be considered.

**Comment by fxd0h** [comment]
Implemented, PR at #499. What a record looks like, straight from `mflux-capabilities --command mflux-generate-z-image-turbo` on that branch (trimmed to three options):

```json
{
  "command": "mflux-generate-z-image-turbo",
  "coverage": "full",
  "options": [
    {
      "flag": "--seed",
      "type": "int",
      "parser_default": null,
      "status": "honored"
    },
    {
      "flag": "--guidance",
      "type": "float",
      "parser_default": null,
      "status": "ignored",
      "reason": "Z-Image Turbo is guidance-distilled; guidance is forced to 0.0."
    },
    {
      "flag": "--negative-prompt",
      "type": "str",
      "parser_default": "",
      "status": "ignored",
      "reason": "CFG is disabled on Z-Image Turbo, so the negative prompt is never encoded."
    }
  ]
}
```

The design goal was that nothing in it can rot: commands come from the installed entry points, options from each CLI's live parser, and the ignored/conditional classification is the same constant the runtime warnings read, so the dump and the warnings cannot disagree. @ianscrivener your two angles are covered: `--format markdown` emits the human/llms.txt-style rendering from the same data, and the JSON is what an Agent Skill or MCP server would consume directly.
